src/engine/NetworkLevelReasoner.cpp

```diff
@@ -1151,6 +1155,93 @@ void NetworkLevelReasoner::log( const String &message )
         printf( "%s", message.ascii() );
 }
 
+bool NetworkLevelReasoner::functionTypeSupported( PiecewiseLinearFunctionType type )
+{
+    if ( type == PiecewiseLinearFunctionType::RELU )
+        return true;
+
+    if ( type == PiecewiseLinearFunctionType::ABSOLUTE_VALUE )
+        return true;
+
+    return false;
+}
+
+void NetworkLevelReasoner::dumpTopology() const
+{
+    printf( "Number of layers: %u.\n", _numberOfLayers );
+    for ( unsigned i = 0; i < _numberOfLayers; ++i )
+        printf( "\t%u\n", _layerSizes[i] );
+
+    for ( unsigned i = 0; i < _numberOfLayers; ++i )
+    {
+        unsigned layerSize = _layerSizes[i];
+        printf( "\nDumping info for layer %u:\n", i );
+        printf( "\tNeurons:\n" );
+        for ( unsigned j = 0; j < layerSize; ++j )
+        {
+            printf( "\t\t" );
+            Index index( i, j );
+            if ( _indexToWeightedSumVariable.exists( index ) )
+                printf( "%4u ", _indexToWeightedSumVariable[index] );
+            else
+                printf( "   " );
+
+            printf( "--> " );
+
+            if ( _indexToActivationResultVariable.exists( index ) )
+                printf( "%4u ", _indexToActivationResultVariable[index] );
+
+            if ( _neuronToActivationFunction.exists( index ) )
+            {
+                switch ( _neuronToActivationFunction[index] )
+                {
+                case PiecewiseLinearFunctionType::RELU:
+                    printf( "   (ReLU)" );
+                    break;
+
+                case PiecewiseLinearFunctionType::ABSOLUTE_VALUE:
+                    printf( "   (Absolute Value)" );
+                    break;
+
+                default:
+                    printf( "   (Unknown)" );
+                    break;
+                }
+            }
+
+            printf( "\n" );
+        }
+
+        if ( i > 0 )
+        {
+            printf( "\n\tEquations:\n" );
+            for ( unsigned j = 0; j < layerSize; ++j )
+            {
+                printf( "\t\tx%u = ", _indexToWeightedSumVariable[Index( i, j )]);
+                for ( unsigned k = 0; k < _layerSizes[i-1]; ++k )
+                {
+                    double weight = _weights[i-1][k * layerSize + j];
+                    if ( FloatUtils::isZero( weight ) )
+                        continue;
+
+                    if ( weight < 0 )
+                        printf( " - %.5lfx%u", -weight, _indexToActivationResultVariable[Index( i - 1, k )] );
+                    else
+                        printf( " + %.5lfx%u", weight, _indexToActivationResultVariable[Index( i - 1, k )] );
```

Comment:
If you do not care for the space between the sign and the value, you can add the + flag in the formatting string " %+.5lfx%u"  to print out the sign for both positive and negative numbers, which would eliminate the need for the if statement.

Reply:
Done